planner_add: command enhancement

WendyBaiYunwei opened the pull request:
Closes #166 

Current `planner_add` command does not support addition of Co-requisites.
Meanwhile, there lacks a check for null.

These problems respectively makes adding of Co-requisites troublesome, and the programme not secure enough.

Let's make `planner_add` support Co-requisites and add in a check for a `DegreePlanner` object that may return null when a `DegreePlanner` with a certain `Year` and `Semester` is not found.

WendyBaiYunwei committed:
planner_add: implement coreq for planner_add

WendyBaiYunwei committed:
planner_add: error-checking (+1 squashed commits)

Squashed commits:

[31c7d9fa] planner_add: prevent null pointer exception

WendyBaiYunwei committed:
planner_add: prevent null pointer exception

WendyBaiYunwei committed:
planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: change method
Currently, getFilteredDegreePlannerList is used directly in the stream.
This can cause unwanted modification to the filtered list.
Let's change it back to the proper method.

WendyBaiYunwei committed:
planner_add: throw exception when error occurs

WendyBaiYunwei committed:
planner_add: make messages private
Currently, messages for planner_add command is public.
Let's make it private as public access modifier is not necessary.

WendyBaiYunwei committed:
planner_add: change null pointer exception to command exception

WendyBaiYunwei committed:
planner_add: change spacing and reorder methods

WendyBaiYunwei committed:
planner_add: make the message of nonexistent plan more specific

WendyBaiYunwei committed:
planner_add: modify implementation
Currently, planner_add adds modules by looping through the moduleList.
This fails to utilize existing method and touches on unFilteredModuleList
Let's change/fix it

WendyBaiYunwei committed:
planner_add: rename variable "newCodeSet" -> "selectedCodeSet"

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -1,6 +1,7 @@
 package seedu.address.logic.commands;
 
 import static java.util.Objects.requireNonNull;
+
```
Remove unnecessary addition of newline.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -37,10 +40,13 @@
+    private static final String MESSAGE_SUCCESS = "Added new module(s) to year %1$s semester %2$s of"
             + " the degree plan: \n%3$s";
-    public static final String MESSAGE_DUPLICATE_CODE = "The module(s) %1$s already exist in the degree plan.";
-    public static final String MESSAGE_MODULE_DOES_NOT_EXIST = "The module(s) %1$s do not exist in the module list.";
```
Why are all the messages made `private`?

WendyBaiYunwei replied:
I have changed the messages back to public. 
But I changed them to private initially because I didn't think they were necessary to be public, since they are used inside the command class only. As we should keep the access modifier as private as possible, I changed them to private.

Creastery replied:
I see. Do note that these message variables need to have `public` access modifiers so that you can refer to them in your unit tests!

WendyBaiYunwei replied:
Anyway, I changed them back to private still because at least for now, these 3 messages will not be used elsewhere. Please just inform me if it is necessary to change back.

Creastery replied:
I think you can set the all the messages as `public`, since you will need to create unit tests to test the logic involving the excluded messages sooner or later.

WendyBaiYunwei replied:
Somehow missed out your comment. I have changed them back. I got something urgent on now and cannot double check again. The tag will still be ongoing. If you feel that the changes meet the review requirement, please directly help me merge. If not, I will come back and double check again. please ignore this message if you have no time now.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -37,10 +40,13 @@
             + " the degree plan: \n%3$s";
-    public static final String MESSAGE_DUPLICATE_CODE = "The module(s) %1$s already exist in the degree plan.";
-    public static final String MESSAGE_MODULE_DOES_NOT_EXIST = "The module(s) %1$s do not exist in the module list.";
+    private static final String COREQ_MESSAGE_SUCCESS = "\nCo-requisite(s) added:\n%1$s";
```
This should be within `MESSAGE_SUCCESS`. 
When no co-requisites are added, simply supply "None" to the format string.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -37,10 +40,13 @@
+    private static final String MESSAGE_DUPLICATE_CODE = "The module(s) %1$s already exists in the degree plan.";
+    private static final String MESSAGE_MODULE_DOES_NOT_EXIST = "The module(s) %1$s does not exist in the module list.";
+    private static final String MESSAGE_PLAN_DOES_NOT_EXIST = "The degree plan of year %1$s and semester"
+            + "%2$s does not exist.";
```
Let's keep variables more concise and consistent.
`MESSAGE_NONEXISTENT_MODULES` and `MESSAGE_NONEXISTENT_DEGREE_PLANNER` are more intuitive.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -79,13 +87,31 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+        Set<Code> selectedCodeSet = new HashSet<>(selectedDegreePlanner.getCodes());
+        Set<Code> coreqAdded = new HashSet<>();
+        selectedCodeSet.addAll(codesToAdd);
+        for (Code codeToAdd :codesToAdd) {
```
Please add space after `:`.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -79,13 +87,31 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+            selectedCodeSet.add(codeToAdd);
+            //adds Co-requisite(s)
+            Module module = model.getModuleByCode(codeToAdd);
+            if (module.getCorequisites().size() > 0) {
```
This if statement check is redundant since `addAll()` does nothing on an empty set anyways.

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -79,13 +87,31 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+            }
+        }
+
+        coreqAdded.removeAll(codesToAdd);
```
Isn't `coreqAdded` going to be equivalent to `selectedCodeSet.removeAll(codesToAdd)`?
If so, you could just do:
```java
Set<Code> coreqAdded = new HashSet<>(selectedCodeSet);
coreqAdded.removeAll(codesToAdd);
```

WendyBaiYunwei replied:
Sorry after changing I realised that perhaps you misunderstood the code. selectedCodeSet contains all the codes in the selected degree planner. When directly adding in all the selectedCodeSet, all the codes that are not Co-req themselves will be added in as well, which is not what we want.

Creastery replied:
Oh okay. In that case, then you can simplify:

https://github.com/CS2113-AY1819S2-T09-1/main/blob/7a97f009b683f2a443190981801685769f8e3886/src/main/java/seedu/address/logic/commands/PlannerAddCommand.java#L90-L103

into:

```java
Set<Code> selectedCodeSet = new HashSet<>(selectedDegreePlanner.getCodes());
selectedCodeSet.addAll(codesToAdd);
codesToAdd.stream().map(model::getModuleByCode).map(Module::getCorequisites).forEach(selectedCodeSet::addAll);

Set<Code> coreqAdded = new HashSet<>(selectedCodeSet);
coreqAdded.removeAll(codesToAdd);
```

WendyBaiYunwei committed:
planner_add: adjustment according to review 1

WendyBaiYunwei committed:
planner_add: change access modifier for messages

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

Creastery commented on src/main/java/seedu/address/logic/commands/PlannerAddCommand.java:
```diff
@@ -76,16 +82,31 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
-        DegreePlanner editedDegreePlanner = new DegreePlanner(yearToAddTo, semesterToAddTo, newCodeSet);
+        DegreePlanner editedDegreePlanner = new DegreePlanner(yearToAddTo, semesterToAddTo, selectedCodeSet);
         model.setDegreePlanner(selectedDegreePlanner, editedDegreePlanner);
         model.commitAddressBook();
```
You are updating only 1 degree planner here.
If the co-requisites you are adding over to this year & semester already exists in other years & semesters, there will be duplicate entries since the existing `DegreePlanner` for the other years & semesters are not updated.

WendyBaiYunwei replied:
Upon reviewing the code just now, I also realised the same problem. To resolve the issue, I threw a command exception and asked the users to move/remove the code(s) involved in order to proceed.

WendyBaiYunwei committed:
planner_add: implement duplicate coreq checking
Currently, as users can edit coreq, they may add duplicate coreq into the degree plan.
Let's implement a check for this and add in valid coreq only.

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
planner_add: improve comments

WendyBaiYunwei committed:
planner_add: adjust comment

WendyBaiYunwei committed:
planner_add: improve code quality

WendyBaiYunwei committed:
planner_add: improve code quality (+1 squashed commits)

Squashed commits:

[d55971a8] planner_add: improve message

WendyBaiYunwei committed:
planner_add: minor enhancement

WendyBaiYunwei committed:
planner_add: downgrade planner_add

WendyBaiYunwei committed:
planner_add: add test

WendyBaiYunwei committed:
planner_add: fix test

WendyBaiYunwei committed:
planner_add: fix test

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
planner_add: fix errors

WendyBaiYunwei committed:
planner_add: resolve merge conflicts

WendyBaiYunwei committed:
planner_add: remove extra imports

WendyBaiYunwei committed:
planner_add: fix test

WendyBaiYunwei committed:
planner_add: resume multiple modules

WendyBaiYunwei committed:
planner_add: fix tests

WendyBaiYunwei committed:
planner_add: fix test

WendyBaiYunwei committed:
planner_add: replace wrongly deleted lines

WendyBaiYunwei commented:
The code for dealing with coreqs left out is maintained. though its function will be invisible in the current version, it can help with future modification of the application.

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerAddCommand.java:
```diff
@@ -39,9 +41,14 @@
+    public static final String MESSAGE_DUPLICATE_CODE = "The module(s) %1$s already exists in the degree plan.";
+    public static final String MESSAGE_INVALID_COREQ = "The Co-requisite(s) %1$s of module(s) %2$s already exists"
+            + " in a different year and semester of the degree plan.\nModules that are Co-requisites to each other"
+            + " have to be in the same year and semester of the degree plan.";
```
`Co-requisites(s)` should be in lowercase.
`Modules that are Co-requisites to each other have to..` should be `Co-requisites modules have to...`

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerAddCommand.java:
```diff
@@ -39,9 +41,14 @@
+            + " have to be in the same year and semester of the degree plan.";
+    public static final String MESSAGE_NONEXISTENT_MODULES = "The module(s) %1$s does not exist in the module list.";
+    public static final String MESSAGE_NONEXISTENT_DEGREE_PLANNER = "The degree plan of year %1$s and semester"
+            + "%2$s does not exist.";
```
Should be:
```
Year %1$s Semester %2$s does not exists in the degree plan!
```

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerAddCommand.java:
```diff
@@ -51,7 +58,6 @@
      */
     public PlannerAddCommand(Year year, Semester semester, Set<Code> codes) {
         requireAllNonNull(year, semester, codes);
-
```
Please add back the line here.
The line is used to separate pre-condition(s) from the remaining code.

Creastery commented on src/test/java/pwe/planner/logic/parser/PlannerAddCommandParserTest.java:
```diff
@@ -0,0 +1,105 @@
+
+        // invalid code
+        assertParseFailure(parser, " " + PREFIX_YEAR + "1 " + PREFIX_SEMESTER + "1 "
+                + PREFIX_CODE + null, Code.MESSAGE_CONSTRAINTS);
```
This test doesn't make sense.

WendyBaiYunwei committed:
planner_add: amend according to review

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
planner_add: minor enhancements

WendyBaiYunwei committed:
planner_add: minor enhancement

WendyBaiYunwei committed:
planner_add: minor enhancement

WendyBaiYunwei committed:
planner_add: minor enhancement

WendyBaiYunwei committed:
planner_add: replace newline with space in message

WendyBaiYunwei committed:
planner_add: change format of message

WendyBaiYunwei committed:
planner_add: change test

WendyBaiYunwei committed:
planner_add: simplify code

WendyBaiYunwei committed:
planner_add: minor change

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerAddCommand.java:
```diff
@@ -63,29 +70,69 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+        Set<Code> nonExistentModuleCodes = codesToAdd.stream()
+                .filter(codesToCheck -> !model.hasModuleCode(codesToCheck)).collect(Collectors.toSet());
+        String nonExistentModuleString = nonExistentModuleCodes.stream().map(Code::toString)
+                .collect(Collectors.joining(", "));
```
This should be declared within the `if` statement.

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerAddCommand.java:
```diff
@@ -63,29 +70,69 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+                .collect(Collectors.toSet());
+        // Converts a set to a string to remove the brackets of set.
+        String duplicatePlannerCodeString = duplicatePlannerCodes.stream().map(Code::toString)
+                .collect(Collectors.joining(", "));
```
This should be declared within the `if` statement.

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerAddCommand.java:
```diff
@@ -63,29 +70,69 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+            // Returns the invalid duplicate co-requisite(s) that exists in a different section of the degree plan.
+            invalidDuplicateCoreqs.removeAll(selectedDegreePlanner.getCodes());
+            if (invalidDuplicateCoreqs.size() > 0) {
+                throw new CommandException(String.format(MESSAGE_INVALID_COREQ, invalidDuplicateCoreqs, codesToAdd));
```
You're lacking `.stream().map(Code::toString).collect(Collectors.joining(", "))` for `invalidDuplicateCoreqs` and `codesToAdd` here.

WendyBaiYunwei committed:
planner_add: change according to review

WendyBaiYunwei committed:
planner_add: enhance test

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

WendyBaiYunwei committed:
Merge branch 'master' into planner_add_new

# Conflicts:
#	src/test/java/pwe/planner/logic/parser/CommandParserTest.java

WendyBaiYunwei committed:
planner_suggest: resolve conflicts

WendyBaiYunwei committed:
Merge branch 'planner_add_new' of https://github.com/WendyBaiYunwei/main into planner_add_new

# Conflicts:
#	src/test/java/pwe/planner/logic/parser/CommandParserTest.java

WendyBaiYunwei committed:
planner_add: add in support for semester

WendyBaiYunwei committed:
planner_add: fix style

WendyBaiYunwei committed:
planner_add: enhance messages

WendyBaiYunwei committed:
planner_add: modify message

WendyBaiYunwei committed:
planner_add: amend message

Creastery reviewed (approved):
LGTM. Approving and merging.
